### PR TITLE
phactory: read authority set change from header

### PR DESCRIPTION
**Pull request by nanometerzhu**

### CODE REVIEW

File: crates/phactory/api/src/blocks.rs [hunk 1]
```diff
@@ -157,3 +157,16 @@ pub mod compat {
...
+    header: &BlockHeader,
+) -> Option<ScheduledChange<u32>> {
+    let filter_log = |log: ConsensusLog<u32>| match log {
+        ConsensusLog::ScheduledChange(change) => Some(change),
+        _ => None,
+    };
+
+    // find the first consensus digest with the right ID which converts to
+    // the right kind of consensus log.
+    header.digest.convert_first(|l| l.consensus_try_to(&GRANDPA_ENGINE_ID).and_then(filter_log))
```
Comment (kvinwang) [inline crates/phactory/api/src/blocks.rs:171]: Would this verify the changes are authorized by the old validators?

Comment (nanometerzhu) [reply crates/phactory/api/src/blocks.rs:171]: Basically it's same data as reading from state. If the header.digest was hacked/broken, the hash would change.

https://spec.polkadot.network/chap-state#defn-block-header

The block hash is hash of (parent_hash + number + state_root + extrinsics_root + digest). Old authority set + justification proves the header hash, header hash proves header.digest

Comment (kvinwang) [reply crates/phactory/api/src/blocks.rs:171]: Got it.

**Review by kvinwang — APPROVED** [review]
Much better than the previous implementation. 👍

**Review by h4x3rotab — APPROVED** [review]
Awesome! Looks very clear

**Pull request merged**